config/config.go

```diff
@@ -30,13 +31,37 @@ const (
 
 	EnvFnRegistry = "registry"
 	EnvFnContext  = "context"
+
+	OCI_CLI_AUTH_ENV_VAR            = "OCI_CLI_AUTH"
+	OCI_CLI_AUTH_INSTANCE_PRINCIPAL = "instance_principal"
+	OCI_CLI_AUTH_INSTANCE_OBO_USER  = "instance_obo_user"
 )
 
-var defaultRootConfigContents = &ContextMap{CurrentContext: "", CurrentCliVersion: ""}
-var DefaultContextConfigContents = &ContextMap{
-	ContextProvider:      DefaultProvider,
-	provider.CfgFnAPIURL: defaultLocalAPIURL,
-	EnvFnRegistry:        "",
+var defaultRootConfigContents = &ContextMap{CurrentContext: "default", CurrentCliVersion: Version}
+
+func DefaultContextConfigContents() (contextMap *ContextMap) {
+	ociCliAuth := os.Getenv(OCI_CLI_AUTH_ENV_VAR)
+
+	if ociCliAuth == OCI_CLI_AUTH_INSTANCE_OBO_USER {
+		contextMap = &ContextMap{
+			ContextProvider: fn_go.OracleCSProvider,
+			EnvFnRegistry:   "",
+		}
+	} else if ociCliAuth == OCI_CLI_AUTH_INSTANCE_PRINCIPAL {
+		contextMap = &ContextMap{
+			ContextProvider: fn_go.OracleIPProvider,
+			EnvFnRegistry:   "",
+		}
+	} else {
+		contextMap = &ContextMap{
+			ContextProvider:      fn_go.DefaultProvider,
+			provider.CfgFnAPIURL: defaultLocalAPIURL,
+			EnvFnRegistry:        "",
+		}
+		viper.SetDefault(provider.CfgFnAPIURL, defaultLocalAPIURL)
```

Comment:
How come `provider.CfgFNAPIURL` is only set for the default case (lines 58 and 61)? It's not clear why it's only being done for one (if someone added a new case, how do they decide whether this is necessary?) Maybe add a clarifying comment.

Reply:
So, for the oracle cloudshell and instance principal providers, the default api url (i.e. only if it's not specified in the context) is determined by the provider (it does so by figuring out the region in the environment).
We don't want to write that url onto file, I think, because the fn context file is in the home directory and on cloudshell for e.g., that home directory is shared across regions.

Reply:
Added a comment to clarify.